File: examples/Clojure/lbbroker.clj
(ns lbbroker
  (:require [zeromq.zmq :as zmq]))

;;
;; Least-recently used (LRU) queue device
;; Clients and workers are shown here in-process
;;
;; While this example runs in a single process, that is just to make
;; it easier to start and stop the example. Each thread has its own
;; context and conceptually acts as a separate process.
;;

(defn client [n]
  (fn []
    (let [ctx (zmq/context 1)
          client (zmq/socket ctx :req)]
      (zmq/set-identity client (-> n str .getBytes))
      (zmq/connect client "ipc://frontend.ipc")
      (zmq/send-str client "HELLO")
      (println "Client:" (zmq/receive-str client))
      (.close client)
      (.term ctx))))

(defn worker [n]
  (fn []
    (let [ctx (zmq/context 1)
          worker (zmq/socket ctx :req)]
      (zmq/set-identity worker (-> n str .getBytes))
      (zmq/connect worker "ipc://backend.ipc")
      (zmq/send-str worker "READY")
      (while true
        (let [identity (zmq/receive-str worker)
              _ (zmq/receive worker)
              request (zmq/receive-str worker)]
          (println "Worker:" request)
          (zmq/send-str worker identity zmq/send-more)
          (zmq/send-str worker "" zmq/send-more)
          (zmq/send-str worker "OK")))
      (.close worker)
      (.term ctx))))

(def worker-nbr 3)
(def client-nbr 10)

(defn -main []
  (let [ctx (zmq/context 1)
        frontend (zmq/socket ctx :router)
        backend (zmq/socket ctx :router)]
    (zmq/bind frontend "ipc://frontend.ipc")
    (zmq/bind backend "ipc://backend.ipc")
    (dotimes [i client-nbr]
      (doto (Thread. (client i))
        (.setDaemon true)
        .start))
    (dotimes [i worker-nbr]
      (doto (Thread. (worker i))
        (.setDaemon true)
        .start))
    ;; Logic of LRU loop
    ;; - Poll backend always, frontend only if 1+ worker ready
    ;; - If worker replies, queue worker as ready and forward reply
    ;; to client if necessary
    ;; - If client requests, pop next worker and send request to it
    (let [clients-left (atom client-nbr)
          worker-queue (atom clojure.lang.PersistentQueue/EMPTY)]
      (while (and (not (.isInterrupted (Thread/currentThread)))
                  (> @clients-left 0))
        (let [items (zmq/poller ctx)
              backend-index (zmq/register items backend :pollin)
              frontend-index (if-not (empty? @worker-queue)
                               (zmq/register items frontend :pollin))]
          (zmq/poll items -1)
          ;; Handle worker activity on background
          (if (zmq/check-poller items backend-index :pollin)
            ;; Queue worker address for LRU routing
            (let [worker-id (zmq/receive-str backend)
                  _ (zmq/receive-str backend)
                  client-id (zmq/receive-str backend)]
              (swap! worker-queue conj worker-id)
              (if (not= "READY" client-id)
                (let [_ (zmq/receive backend)
                      reply (zmq/receive-str backend)]
                  (zmq/send-str frontend client-id zmq/send-more)
                  (zmq/send-str frontend "" zmq/send-more)
                  (zmq/send-str frontend reply)
                  (swap! clients-left dec)))))
          (if (and frontend-index (zmq/check-poller items frontend-index :pollin))
            (let [client-id (zmq/receive-str frontend)
                  _ (zmq/receive frontend)
                  request (zmq/receive-str frontend)
                  worker-id (peek @worker-queue)]
              (swap! worker-queue pop)
              (zmq/send-str backend worker-id zmq/send-more)
              (zmq/send-str backend "" zmq/send-more)
              (zmq/send-str backend client-id zmq/send-more)
              (zmq/send-str backend "" zmq/send-more)
              (zmq/send-str backend request))))))
    (.close frontend)
    (.close backend)
    (.term ctx)))

File: examples/C++/lbbroker.cpp
//  Least-recently used (LRU) queue device
//  Clients and workers are shown here in-process
//

#include "zhelpers.hpp"
#include <thread>
#include <queue>

//  Basic request-reply client using REQ socket
//
void receive_empty_message(zmq::socket_t& sock)
{
    std::string empty = s_recv(sock);
    assert(empty.size() == 0);
}

void client_thread(int id) {
    zmq::context_t context(1);
    zmq::socket_t client(context, ZMQ_REQ);

#if (defined (WIN32))
    s_set_id(client, id);
    client.connect("tcp://localhost:5672"); // frontend
#else
    s_set_id(client); // Set a printable identity
    client.connect("ipc://frontend.ipc");
#endif

    //  Send request, get reply
    s_send(client, std::string("HELLO"));
    std::string reply = s_recv(client);
    std::cout << "Client: " << reply << std::endl;
    return;
}

//  Worker using REQ socket to do LRU routing
//
void worker_thread(int id) {
    zmq::context_t context(1);
    zmq::socket_t worker(context, ZMQ_REQ);

#if (defined (WIN32))
    s_set_id(worker, id);
    worker.connect("tcp://localhost:5673"); // backend
#else
    s_set_id(worker);
    worker.connect("ipc://backend.ipc");
#endif

    //  Tell backend we're ready for work
    s_send(worker, std::string("READY"));

    while (1) {
        //  Read and save all frames until we get an empty frame
        //  In this example there is only 1 but it could be more
        std::string address = s_recv(worker);
	receive_empty_message(worker);
        //  Get request, send reply
        std::string request = s_recv(worker);
        std::cout << "Worker: " << request << std::endl;

        s_sendmore(worker, address);
        s_sendmore(worker, std::string(""));
        s_send(worker, std::string("OK"));
    }
    return;
}



int main(int argc, char *argv[])
{

    //  Prepare our context and sockets
    zmq::context_t context(1);
    zmq::socket_t frontend(context, ZMQ_ROUTER);
    zmq::socket_t backend(context, ZMQ_ROUTER);

#if (defined (WIN32))
    frontend.bind("tcp://*:5672"); // frontend
    backend.bind("tcp://*:5673"); // backend
#else
    frontend.bind("ipc://frontend.ipc");
    backend.bind("ipc://backend.ipc");
#endif

    int client_nbr = 0;
    for (; client_nbr < 10; client_nbr++) {
        std::thread t(client_thread, client_nbr);
	t.detach();
    }
    for (int worker_nbr = 0; worker_nbr < 3; worker_nbr++) {
	    std::thread t (worker_thread, worker_nbr);
            t.detach();
    }
    //  Logic of LRU loop
    //  - Poll backend always, frontend only if 1+ worker ready
    //  - If worker replies, queue worker as ready and forward reply
    //    to client if necessary
    //  - If client requests, pop next worker and send request to it
    //
    //  A very simple queue structure with known max size
    std::queue<std::string> worker_queue;

    while (1) {

        //  Initialize poll set
        zmq::pollitem_t items[] = {
                //  Always poll for worker activity on backend
                { backend, 0, ZMQ_POLLIN, 0 },
                //  Poll front-end only if we have available workers
                { frontend, 0, ZMQ_POLLIN, 0 }
        };
        if (worker_queue.size())
            zmq::poll(&items[0], 2, -1);
        else
            zmq::poll(&items[0], 1, -1);

        //  Handle worker activity on backend
        if (items[0].revents & ZMQ_POLLIN) {

            //  Queue worker address for LRU routing
            worker_queue.push(s_recv(backend));
	    receive_empty_message(backend);

            //  Third frame is READY or else a client reply address
            std::string client_addr = s_recv(backend);

            //  If client reply, send rest back to frontend
            if (client_addr.compare("READY") != 0) {
                receive_empty_message(backend);
                std::string reply = s_recv(backend);
                s_sendmore(frontend, client_addr);
                s_sendmore(frontend, std::string(""));
                s_send(frontend, reply);

                if (--client_nbr == 0)
                    break;
            }
        }
        if (items[1].revents & ZMQ_POLLIN) {

            //  Now get next client request, route to LRU worker
            //  Client request is [address][empty][request]
            std::string client_addr = s_recv(frontend);

            {
                std::string empty = s_recv(frontend);
                assert(empty.size() == 0);
            }

            std::string request = s_recv(frontend);

            std::string worker_addr = worker_queue.front();//worker_queue [0];
            worker_queue.pop();

            s_sendmore(backend, worker_addr);
            s_sendmore(backend, std::string(""));
            s_sendmore(backend, client_addr);
            s_sendmore(backend, std::string(""));
            s_send(backend, request);
        }
    }
    return 0;
}
